adam/learner/cross_situational.py

```diff
@@ -0,0 +1,567 @@
+        observed_linguistic_description = (
+            learning_example.linguistic_description.as_token_sequence()
+        )
+
+        self.learn_with_cross_situational(
+            observed_perception_graph, observed_linguistic_description
+        )
+
+    # Re-implement for cross-situational
+    def learn_with_cross_situational(
+        self,
+        observed_perception_graph: PerceptionGraph,
+        observed_linguistic_description: Tuple[str, ...],
+    ) -> None:
+        logging.info(
+            f"Cross-situational learner observing {observed_linguistic_description}"
+        )
+        # The learner’s words are W, meanings are M, their associations are A, and the new
+        # utterance is U = (W_U, M_U).
+        # For every w in W_U
+        # Adding "dummy word" that smoothes out learning for situations where a word shows up
+        # without it's meaning or vice versa. Also smoothes out meanings that show up without a word
+        # in general.
+        words = observed_linguistic_description + tuple("d")
+        meanings = [
+            PerceptionGraphPattern.from_graph(object_).perception_graph_pattern
+            for object_ in self.get_objects_from_perception(observed_perception_graph)
+        ]
+        for word in words:
+            # TODO: pursuit learner hard-coded to ignore determiners
+            # https://github.com/isi-vista/adam/issues/498
+            if word in ("a", "the"):
+                continue
+            if word in self._words_to_number_of_observations:
+                self._words_to_number_of_observations[word] += 1
+            else:
+                self._words_to_number_of_observations[word] = 1
+
+            # If word is not known, go through learning steps
+            if word not in self._lexicon:
+                logging.info(f"Considering '{word}'")
+                # Initialize necessary components if needed
+                if word not in self._words_to_hypotheses_and_scores:
+                    self._words_to_hypotheses_and_scores[word] = {}
+                    self._words_to_hypotheses_and_probability[word] = {}
+                # For each meaning, update the association score for the word meaning pair
+                for meaning in meanings:
+                    for hypothesis in self._words_to_hypotheses_and_scores[word]:
+                        if hypothesis.check_isomorphism(meaning):
+                            # Updating association score as described by the paper below
+                            # assoc(w, m) = pass_assoc(w, m) + a(w|m, U)
+                            # https://onlinelibrary.wiley.com/doi/full/10.1111/j.1551-6709.2010.01104.x (2)
+                            self._words_to_hypotheses_and_scores[word][
+                                hypothesis
+                            ] += self.alignment_probability(word, words, hypothesis)
+                            break
+                    else:
```

Comment:
There seems to be an indentation issue with this clause

Reply:
The `else:` is for `for hypothesis in self._words_to_hypotheses_and_scores[word]:` so the indentation is intended unless there's something I'm not seeing. @denizbeser

Reply:
I see, didn't notice it's for/else.